[CMake] Generate a usable compile_commands.json

TingPing opened the pull request:
#### 5c9dd47913bc1cc71ae7cea6916a1b5eae19b5e3
<pre>
[CMake] Generate a usable compile_commands.json
<a href="https://bugs.webkit.org/show_bug.cgi?id=260218">https://bugs.webkit.org/show_bug.cgi?id=260218</a>

Reviewed by Michael Catanzaro.

When using unified builds CMake generates a compile_commands.json
that isn&apos;t useful by many tools as it lacks much of the sources.

This adds a script to expand UnifiedSources into the full list
of sources. It outputs it to webkit_compile_commands.json.

* Source/cmake/WebKitCommon.cmake:
* Tools/Scripts/rewrite-compile-commands: Added.

Canonical link: <a href="https://commits.webkit.org/266947@main">https://commits.webkit.org/266947@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/94635253a4879c3e8f139be0435a0f678753409b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/15214 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15519 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15879 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16970 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14291 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15354 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/18034 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15617 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16904 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15397 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15838 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12937 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17703 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/13118 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13725 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20688 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/13032 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/14198 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13893 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/17149 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/14463 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14459 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12254 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/15407 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13734 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3929 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3658 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18078 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/15642 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14296 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3740 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

mcatanzaro reviewed (approved):
I would add a --help argument, even though the script is probably not intended to be called manually. Somebody else will thank you for it a few years from now....

TingPing committed:
[CMake] Generate a usable compile_commands.json
https://bugs.webkit.org/show_bug.cgi?id=260218

Reviewed by Michael Catanzaro.

When using unified builds CMake generates a compile_commands.json
that isn't useful by many tools as it lacks much of the sources.

This adds a script to expand UnifiedSources into the full list
of sources. It outputs it to webkit_compile_commands.json.

* Source/cmake/WebKitCommon.cmake:
* Tools/Scripts/rewrite-compile-commands: Added.

Canonical link: https://commits.webkit.org/266947@main

philn commented on Tools/Scripts/rewrite-compile-commands:
```diff
@@ -0,0 +1,92 @@
+                    })
+                    break
+            else:
+                print('Failed to find', include_path)
```
This triggers for many files here, `*MessageReceiver.cpp` that don't exist and files from `WebCore/DerivedSources/` mostly.

Is this OK?

TingPing replied:
Hmm in my local builds it found everything. What was your specific build config (and which commit maybe)?

TingPing replied:
Also was it ran *after* everything was built? You can run `ninja webkit_compile_commands.json` after a build to run it last.

We are kinda in a weird situation where this script technically depends on every generated source, maybe we could somehow actually collect that list, but I don't know a simple solution... So maybe it runs out of order sometimes.

I didn't want to rely on checking for files existing but the rules of where a file ends up didn't have a clear pattern, but maybe some static mapping is possible.

philn replied:
Just doing a clean GTK Debug build triggers the issue here, with sccache enabled, (*edit:* with ASan enabled too) the script runs early on, I suppose before the MessageReceiver.cpp files have been generated.

philn replied:
Actually those MessageReceiver.cpp files are for features we don't enable, such as GPUProcess, so it's expected that they shouldn't exist.

TingPing replied:
> Actually those MessageReceiver.cpp files are for features we don't enable, such as GPUProcess, so it's expected that they shouldn't exist.

The file has to exist in some include path, it is compiled, otherwise the build would fail.
